crates/interpreter/src/opcode.rs

```diff
@@ -276,43 +284,132 @@ impl OpCode {
+    imm_size: u8,
+}
+
+impl fmt::Debug for OpCodeInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("OpCodeInfo")
+            .field("name", &self.name())
+            .field("inputs", &self.inputs())
+            .field("outputs", &self.outputs())
+            .field("not_eof", &self.is_disabled_in_eof())
+            .field("terminating", &self.is_terminating())
+            .field("imm_size", &self.imm_size())
+            .finish()
+    }
 }
 
 impl OpCodeInfo {
+    /// Creates a new opcode info with the given name and default values.
     pub const fn new(name: &'static str) -> Self {
+        assert!(name.len() < 256, "opcode name is too long");
         Self {
-            name,
+            name_ptr: unsafe { NonNull::new_unchecked(name.as_ptr().cast_mut()) },
+            name_len: name.len() as u8,
             inputs: 0,
             outputs: 0,
-            is_eof: true,
-            is_terminating_opcode: false,
-            immediate_size: 0,
+            not_eof: false,
+            terminating: false,
+            imm_size: 0,
+        }
+    }
+
+    /// Returns the opcode name.
+    #[inline]
+    pub const fn name(&self) -> &'static str {
+        // SAFETY: `self.name_*` can only be initialized with a valid `&'static str`.
+        unsafe {
+            // TODO: Use `str::from_raw_parts` when it's stable.
+            let slice = core::slice::from_raw_parts(self.name_ptr.as_ptr(), self.name_len as usize);
+            core::str::from_utf8_unchecked(slice)
         }
     }
 
+    /// Calculates the difference between the number of input and output stack elements.
+    #[inline]
     pub const fn io_diff(&self) -> i16 {
         self.outputs as i16 - self.inputs as i16
     }
+
+    /// Returns the number of input stack elements.
+    #[inline]
+    pub const fn inputs(&self) -> u8 {
+        self.inputs
+    }
+
+    /// Returns the number of output stack elements.
+    #[inline]
+    pub const fn outputs(&self) -> u8 {
+        self.outputs
+    }
+
+    /// Returns whether this opcode is disabled in EOF bytecode.
+    #[inline]
+    pub const fn is_disabled_in_eof(&self) -> bool {
+        self.not_eof
+    }
+
+    /// Returns whether this opcode terminates execution, e.g. `STOP`, `RETURN`, etc.
+    #[inline]
+    pub const fn is_terminating(&self) -> bool {
+        self.terminating
+    }
+
+    /// Returns the size of the immediate value in bytes.
+    #[inline]
+    pub const fn imm_size(&self) -> u8 {
+        self.imm_size
+    }
+
+    /// Sets the EOF flag to false.
+    #[inline]
+    pub const fn not_eof(mut self) -> Self {
+        self.not_eof = true;
+        self
+    }
+
+    /// Sets the immediate bytes number.
+    ///
+    /// RJUMPV is special case where the bytes len is depending on bytecode value,
+    /// for RJUMPV size will be set to one byte while minimum is two.
+    #[inline]
+    pub const fn set_imm_size(mut self, n: u8) -> Self {
```

Comment:
Would leave those three as standalone functions as their usage (at least for now) should be only related to `opcodes` macro.

Reply:
They're used in builder-style instantiation, so they make more sense as methods. I can however make them private if you want.

Reply:
I would leave them as standalone, as in that way we dont clash with naming.